1. Загрузить данные из файла “ECDCCases.csv”.

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('ECDCCases.csv')
print(df.shape)
df.head(3)

(61904, 12)


dateRep  day  month  year  cases  deaths countriesAndTerritories geoId  \
0  14/12/2020   14     12  2020    746       6             Afghanistan    AF   
1  13/12/2020   13     12  2020    298       9             Afghanistan    AF   
2  12/12/2020   12     12  2020    113      11             Afghanistan    AF   

  countryterritoryCode  popData2019 continentExp  \
0                  AFG   38041757.0         Asia   
1                  AFG   38041757.0         Asia   
2                  AFG   38041757.0         Asia   

   Cumulative_number_for_14_days_of_COVID-19_cases_per_100000  
0                                           9.013779           
1                                           7.052776           
2                                           6.868768

2. Проверить в данных наличие пропущенных значений. Вывести 
количество пропущенных значений в процентах. Удалить два признака, в 
которых больше всех пропущенных значений. Для оставшихся признаков 
обработать пропуски: для категориального признака использовать заполнение 
значением по умолчанию (например, «other»), для числового признака 
использовать заполнение медианным значением. Показать, что пропусков 
больше в данных нет.

In [3]:
for column in df.columns:
    missing_values = np.mean(df[column].isna() * 100)
    print(f"{column}: {round(missing_values, 2)}%")

dateRep: 0.0%
day: 0.0%
month: 0.0%
year: 0.0%
cases: 0.0%
deaths: 0.0%
countriesAndTerritories: 0.0%
geoId: 0.44%
countryterritoryCode: 0.2%
popData2019: 0.2%
continentExp: 0.0%
Cumulative_number_for_14_days_of_COVID-19_cases_per_100000: 4.65%


In [4]:
df.drop(['geoId', 'Cumulative_number_for_14_days_of_COVID-19_cases_per_100000'], axis=1, inplace=True)

In [5]:
df.dtypes

dateRep                     object
day                          int64
month                        int64
year                         int64
cases                        int64
deaths                       int64
countriesAndTerritories     object
countryterritoryCode        object
popData2019                float64
continentExp                object
dtype: object

In [6]:
# countryterritoryCode - category
# popData2019 - float
df.countryterritoryCode.fillna('other', inplace=True)
df.popData2019.fillna(df.popData2019.median(), inplace=True)

In [7]:
df.isna().sum()

dateRep                    0
day                        0
month                      0
year                       0
cases                      0
deaths                     0
countriesAndTerritories    0
countryterritoryCode       0
popData2019                0
continentExp               0
dtype: int64

3. Посмотреть статистику по данным, используя describe(). Сделать 
выводы о том, какие признаки содержат выбросы. Посмотреть, для каких 
стран количество смертей в день превысило 3000 и сколько таких дней было.

In [8]:
df.describe()

day         month          year          cases        deaths  \
count  61904.000000  61904.000000  61904.000000   61904.000000  61904.000000   
mean      15.629232      7.067104   2019.998918    1155.079026     26.053987   
std        8.841624      2.954816      0.032881    6779.010824    131.222948   
min        1.000000      1.000000   2019.000000   -8261.000000  -1918.000000   
25%        8.000000      5.000000   2020.000000       0.000000      0.000000   
50%       15.000000      7.000000   2020.000000      15.000000      0.000000   
75%       23.000000     10.000000   2020.000000     273.000000      4.000000   
max       31.000000     12.000000   2020.000000  234633.000000   4928.000000   

        popData2019  
count  6.190400e+04  
mean   4.091909e+07  
std    1.529798e+08  
min    8.150000e+02  
25%    1.324820e+06  
50%    7.169456e+06  
75%    2.851583e+07  
max    1.433784e+09

Данные, которые содержать выбросы (названия колонок):
- cases
- deaths
- day

In [9]:
deaths_above_3k = df.loc[df.deaths > 3000][['countryterritoryCode', 'dateRep', 'deaths']]
deaths_above_3k

countryterritoryCode     dateRep  deaths
2118                   ARG  02/10/2020    3351
16908                  ECU  07/09/2020    3800
37038                  MEX  09/10/2020    3013
44888                  PER  14/08/2020    3935
44909                  PER  24/07/2020    3887
59007                  USA  12/12/2020    3343
59009                  USA  10/12/2020    3124
59016                  USA  03/12/2020    3190
59239                  USA  24/04/2020    3179
59245                  USA  18/04/2020    3770
59247                  USA  16/04/2020    4928

In [10]:
print(f'Страны, для которых количество смертей в день превысило 3000: {deaths_above_3k.countryterritoryCode.unique()}')
print(f'Всего таких дней: {len(deaths_above_3k)}')

Страны, для которых количество смертей в день превысило 3000: ['ARG' 'ECU' 'MEX' 'PER' 'USA']
Всего таких дней: 11


4. Найти дублирование данных. Удалить дубликаты.

In [11]:
df.duplicated().sum()

4

In [12]:
df.drop_duplicates(inplace=True)

5. Загрузить данные из файла “bmi.csv”. Взять оттуда две выборки. Одна 
выборка – это индекс массы тела людей c региона northwest, вторая выборка –
это индекс массы тела людей с региона southwest. Сравнить средние значения 
этих выборок, используя t-критерий Стьюдента. Предварительно проверить 
выборки на нормальность (критерий Шопиро-Уилка) и на гомогенность 
дисперсии (критерий Бартлетта).

In [13]:
bmi = pd.read_csv('bmi.csv')
print(bmi.shape)
bmi.head(3)

(650, 2)


bmi     region
0  22.705  northwest
1  28.880  northwest
2  27.740  northwest

In [14]:
bmi_northwest = bmi.loc[bmi.region == 'northwest']
bmi_southwest = bmi.loc[bmi.region == 'southwest']

In [15]:
print(bmi_northwest.shape[0])
print(bmi_southwest.shape[0])

325
325


In [16]:
# Проверка распределений на нормальность (критерий Шапиро-Уилка)
import scipy.stats as st

res1 = st.shapiro(bmi_northwest.iloc[:, 0])
res2 = st.shapiro(bmi_southwest.iloc[:, 0])
print(f'{res1} \n{res2}')

(0.9954643249511719, 0.46545225381851196) 
(0.994927167892456, 0.3630009591579437)


Тест дает p-значение выше 0.05, следовательно, выборки имеют
нормальное распределение.

In [17]:
# Проверка гомогенности дисперсии. Критерий Бартлетта
res = st.bartlett(bmi_northwest.iloc[:, 0], bmi_southwest.iloc[:, 0])
print(res)

BartlettResult(statistic=3.4000745256463825, pvalue=0.06519347353580048)


Уровень превышает 0.05, следовательно, дисперсии выборок примерно
одинаковы.

In [19]:
# t-критерий Стьюдента
t_res = st.ttest_ind(bmi_northwest.iloc[:, 0], bmi_southwest.iloc[:, 0])
print(t_res)

Ttest_indResult(statistic=-3.2844171500398582, pvalue=0.001076958496307695)


p-значение ниже 0.05, следовательно нулевая гипотеза
отвергается, выборки принадлежат к разным генеральным совокупностям и их
средние значения различны.

6. Кубик бросили 600 раз, получили следующие результаты: (см. pdf).
С помощью критерия Хи-квадрат проверить, является ли полученное
распределение равномерным. Использовать функцию scipy.stats.chisquare().

In [23]:
data_cube = pd.DataFrame({'N': [1, 2, 3, 4, 5, 6],
                     'Observed': [97, 98, 109, 95, 97, 104],
                     'Expected': [100, 100, 100, 100, 100, 100]})
data_cube

N  Observed  Expected
0  1        97       100
1  2        98       100
2  3       109       100
3  4        95       100
4  5        97       100
5  6       104       100

In [24]:
st.chisquare(data_cube.Observed, data_cube.Expected)

Power_divergenceResult(statistic=1.44, pvalue=0.9198882077437889)

p-value больше, чем 0,05. Исходя из этого мы видим, что распределение является равномерным.

7. С помощью критерия Хи-квадрат проверить, являются ли переменные 
зависимыми.
Создать датафрейм, используя следующий код:

In [25]:
data = pd.DataFrame({'Женат': [89,17,11,43,22,1],
                     'Гражданский брак': [80,22,20,35,6,4],
                     'Не состоит в отношениях': [35,44,35,6,8,22]})

data.index = ['Полный рабочий день','Частичная занятость','Временно не работает',
              'На домохозяйстве','На пенсии','Учёба']

data

Женат  Гражданский брак  Не состоит в отношениях
Полный рабочий день      89                80                       35
Частичная занятость      17                22                       44
Временно не работает     11                20                       35
На домохозяйстве         43                35                        6
На пенсии                22                 6                        8
Учёба                     1                 4                       22

Использовать функцию scipy.stats.chi2_contingency().
Влияет ли семейное положение на занятость?

In [27]:
st.chi2_contingency(data)[:3]

(122.29654948595365, 1.7291616900960234e-21, 10)

**Интерпретация:**
    
Статистика теста хи-квадрат: 122.3
    
р-значение: 1.73e-21
    
Степени свободы: 10 (рассчитывается как (rows - 1)* (columns - 1))
    
p-значение меньше 0.05, семейное положение влияет на занятость.